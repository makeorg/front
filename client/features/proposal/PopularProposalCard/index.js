// @flow
import React from 'react';
import { i18n } from 'Shared/i18n';
import { isInProgress } from 'Shared/helpers/date';
import { type ProposalType } from 'Shared/types/proposal';
import {
  ProposalStyle,
  ProposalInnerStyle,
} from 'Client/ui/Elements/ProposalCardElements';
import { ProposalAuthorElement } from 'Client/ui/Proposal/AuthorElement';
import { AuthorWrapperStyle } from 'Client/ui/Proposal/AuthorElement/Styled';
import { ScreenReaderItemStyle } from 'Client/ui/Elements/AccessibilityElements';
import { Vote } from 'Client/features/vote';
import { DetailledVoteResults } from 'Client/ui/Proposal/DetailledVoteResults';
import { getProposalLink } from 'Shared/helpers/url';
import { useSelector } from 'react-redux';
import { type StateRoot } from 'Shared/store/types';
import {
  PopularProposalWrapperStyle,
  PopularProposalHeader,
  PopularProposalTagStyle,
} from './style';

type Props = {
  /** Object with all proposal's properties */
  proposal: ProposalType,
  /** Proposal's position in list */
  position: number,
  /** Size of proposals list */
  size: number,
};

export const PopularProposalCard = ({
  proposal,
  position = 0,
  size = 0,
}: Props) => {
  const { country } = useSelector((state: StateRoot) => state.appConfig);
  const canVote = isInProgress(proposal.question);
  return (
    <PopularProposalWrapperStyle aria-posinset={position} aria-setsize={size}>
      <PopularProposalHeader aria-hidden>
        {`#${position}`}
      </PopularProposalHeader>
      <AuthorWrapperStyle>
        <ProposalAuthorElement proposal={proposal} withAvatar />
      </AuthorWrapperStyle>
      <ProposalInnerStyle>
        <ScreenReaderItemStyle>
          {i18n.t('proposal_card.content')}
        </ScreenReaderItemStyle>
        <ProposalStyle
          id={`proposal_content_${position}`}
          to={getProposalLink(
            country,
            proposal.slug,
            proposal.id,
            proposal.slug
          )}
        >
          {proposal.content}
        </ProposalStyle>
        {canVote ? (
          <Vote
            proposal={proposal}
            votes={proposal.votes}
            proposalKey={proposal.proposalKey}
            index={position}
          />
        ) : (
          <DetailledVoteResults
            votes={proposal.votes}
            proposalId={proposal.id}
          />
        )}
      </ProposalInnerStyle>
      {proposal.selectedStakeTag && proposal.selectedStakeTag.display && (
        <PopularProposalTagStyle>
          <ScreenReaderItemStyle
            dangerouslySetInnerHTML={{
              __html: i18n.t('consultation.tags.proposal_list'),
            }}
          />
          {proposal.selectedStakeTag.label}
        </PopularProposalTagStyle>
      )}
    </PopularProposalWrapperStyle>
  );
};
