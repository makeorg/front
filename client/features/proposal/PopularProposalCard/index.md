##### PopularProposalCard Example :

```jsx
import { proposalTypeFixture } from '../../../../shared/types/__fixtures__/proposal.fixture.js';
import { PopularProposalCard } from './index';

<ol
  style={{
    width: '100% ',
    height: '440px',
    listStyle: 'none',
    padding: '15px',
    background: '#f2f2f2',
  }}
>
  <PopularProposalCard proposal={proposalTypeFixture} />
</ol>;
```
