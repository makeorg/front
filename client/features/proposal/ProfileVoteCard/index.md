##### ProfileVoteCard Example :
```jsx
import { proposalTypeFixture } from '../../../../shared/types/__fixtures__/proposal.fixture.js';
import { organisationTypeFixture } from '../../../../shared/types/__fixtures__/organisation.fixture.js';
import { ProfileVoteCard } from './index';

<ul style={{listStyle: 'none', padding: '15px', background: '#f2f2f2'}}>
  <ProfileVoteCard 
    voteKey='agree'
    proposal={proposalTypeFixture}
    organisation={organisationTypeFixture}
  />
</ul>
```