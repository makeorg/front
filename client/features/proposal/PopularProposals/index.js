// @flow
import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { TileWithTitle } from 'Client/ui/Elements/TileWithTitle';
import { type Question } from 'Shared/types/question';
import { type TypeSliderParams } from 'Shared/types/views';
import { i18n } from 'Shared/i18n';
import { fetchPopularProposals } from 'Shared/store/actions/question';
import {
  type StateRoot,
  type PopularProposals as PopularProposalsType,
} from 'Shared/store/types';
import { PopularProposalCard } from 'Client/features/proposal/PopularProposalCard';
import { useSlider } from 'Client/hooks/useSlider';
import { Breakpoints } from 'Client/app/assets/vars/Breakpoints';
import { SvgArrowLeft, SvgArrowRight } from 'Client/ui/Svg/elements';
import { FlexElementStyle } from 'Client/ui/Elements/FlexElements';
import { HiddenItemStyle } from 'Client/ui/Elements/HiddenElements';
import { GliderStylesheet } from 'Client/app/assets/css-in-js/GliderStyle';
import { ProposalsSliderWrapperStyle } from 'Client/features/homepage/Proposals/Styled';
import { BackgroundColors } from 'Client/app/assets/vars/Colors';
import { useMobile } from 'Client/hooks/useMedia';
import {
  PopularProposalsSliderListItemStyle,
  PopularProposalsSliderListStyle,
  ArrowStyle,
} from './style';

type Props = {
  question: Question,
};

const sliderName = 'popularProposals';
const sliderParams: TypeSliderParams = {
  slidesToShow: 1.5,
  responsive: [
    {
      breakpoint: Breakpoints.Tablet,
      settings: {
        slidesToShow: 2.5,
      },
    },
  ],
  draggable: true,
  interactiveChildren: {
    links: true,
    buttons: true,
  },
  arrows: {
    prev: `.${sliderName}.glider-prev`,
    next: `.${sliderName}.glider-next`,
  },
};

export const PopularProposals = ({ question }: Props) => {
  const sliderRef = useRef();
  const isMobile = useMobile();
  const dispatch = useDispatch();
  const proposals: PopularProposalsType = useSelector(
    (state: StateRoot) => state.questions[question.slug].popularProposals
  );

  useEffect(() => {
    dispatch(fetchPopularProposals(question.questionId, question.slug));
  }, [dispatch, question.questionId]);

  const hasProposals = proposals && proposals.results.length > 0;

  useSlider(sliderRef, sliderParams, hasProposals);

  if (!hasProposals) {
    return null;
  }

  return (
    <TileWithTitle
      title={i18n.t('consultation.popular_proposals.title', {
        count: proposals.results.length,
      })}
      style={{
        backgroundColor: BackgroundColors.ExtraLightGrey,
        marginTop: '15px',
      }}
    >
      <ProposalsSliderWrapperStyle className={`${sliderName} glider-contain`}>
        <GliderStylesheet />
        <FlexElementStyle as={isMobile ? HiddenItemStyle : FlexElementStyle}>
          <ArrowStyle
            className={`${sliderName} glider-prev`}
            aria-label={i18n.t('common.slider.previous')}
            aria-controls="glide_translator"
            direction="left"
          >
            <SvgArrowLeft aria-hidden />
          </ArrowStyle>
          <ArrowStyle
            className={`${sliderName} glider-next`}
            aria-label={i18n.t('common.slider.next')}
            aria-controls="glide_translator"
            direction="right"
          >
            <SvgArrowRight aria-hidden />
          </ArrowStyle>
        </FlexElementStyle>
        <div className={`${sliderName} glider`} ref={sliderRef}>
          <PopularProposalsSliderListStyle
            className={`${sliderName} glider-track`}
          >
            {proposals.results.map((proposal, index) => (
              <PopularProposalsSliderListItemStyle
                key={proposal.id}
                className={sliderName}
              >
                <PopularProposalCard
                  proposal={proposal}
                  position={index + 1}
                  size={proposals.results.length}
                />
              </PopularProposalsSliderListItemStyle>
            ))}
          </PopularProposalsSliderListStyle>
        </div>
      </ProposalsSliderWrapperStyle>
    </TileWithTitle>
  );
};
