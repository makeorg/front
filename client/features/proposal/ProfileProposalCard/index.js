// @flow
import React from 'react';
import { type Proposal as TypeProposal } from 'Shared/types/proposal';
import { getProposalLink, getConsultationLink } from 'Shared/helpers/url';
import { DetailledVoteResults } from 'Client/ui/Proposal/DetailledVoteResults';
import { ProposalAuthorElement } from 'Client/ui/Proposal/AuthorElement';
import { ProposalFooterWithQuestionElement } from 'Client/ui/Proposal/FooterElement';
import { ProfileProposalCardStyle, ProposalStyle } from './Styled';

type Props = {
  /** Object with all proposal's properties */
  proposal: TypeProposal,
  /** Proposal's position in list */
  position: number,
  /** Size of proposals list */
  size: number,
  /** Optional display for proposal's status */
  withStatus?: boolean,
};

export const ProfileProposalCard = (props: Props) => {
  const { proposal, position, size } = props;
  const { author, question } = proposal;
  const formattedProposalStatus = proposal.status.toLowerCase();
  const isProposalAccepted = formattedProposalStatus === 'accepted';

  return (
    <ProfileProposalCardStyle
      role="article"
      aria-posinset={position}
      aria-setsize={size}
      className={`proposal-${formattedProposalStatus}`}
    >
      <ProposalAuthorElement
        author={author}
        country={proposal.country}
        language={proposal.language}
        createdAt={proposal.createdAt}
        withAvatar
        formattedProposalStatus={formattedProposalStatus}
      />
      <ProposalStyle
        id={`proposal_content_${position}`}
        {...(isProposalAccepted
          ? {
              href: getProposalLink(
                proposal.country,
                proposal.language,
                question.slug,
                proposal.id,
                proposal.slug
              ),
            }
          : { as: 'p' })}
      >
        {proposal.content}
      </ProposalStyle>
      {isProposalAccepted && (
        <DetailledVoteResults votes={proposal.votes} proposalId={proposal.id} />
      )}
      <ProposalFooterWithQuestionElement
        question={question}
        consultationLink={getConsultationLink(
          proposal.country,
          proposal.language,
          question.slug
        )}
        isProposalAccepted={isProposalAccepted}
      />
    </ProfileProposalCardStyle>
  );
};

ProfileProposalCard.defaultPropTypes = {
  withStatus: false,
};
