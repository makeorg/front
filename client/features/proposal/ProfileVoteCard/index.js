// @flow

import React from 'react';
import { i18n } from 'Shared/i18n';
import { type Proposal as TypeProposal } from 'Shared/types/proposal';
import { type Organisation as TypeOrganisation } from 'Shared/types/organisation';
import {
  getProposalLink,
  getOrganisationProfileLink,
} from 'Shared/helpers/url';
import { voteStaticParams } from 'Shared/constants/vote';
import { RedLinkStyle } from 'Client/ui/Elements/LinkElements';
import { SvgCheckedSymbol, SvgThumbsUp } from 'Client/ui/Svg/elements';
import { VoteResultElement } from 'Client/ui/Proposal/VoteResultElement';
import { ProposalFooterWithTagElement } from 'Client/ui/Proposal/FooterElement/ProposalWithTag';
import { ProposalAuthorElement } from 'Client/ui/Proposal/AuthorElement';
import { TextColors } from 'Client/app/assets/vars/Colors';
import {
  ProposalCardStyle,
  ProposalStyle,
  ProposalInnerStyle,
} from 'Client/ui/Elements/ProposalCardElements';
import { AuthorWrapperStyle } from 'Client/ui/Proposal/AuthorElement/Styled';
import { ButtonIconWrapperStyle } from 'Client/ui/Elements/Vote/Styled';
import {
  ProfileVoteCardStyle,
  ProfileVoteWrapperStyle,
  ProfileVoteTitleStyle,
  ProfileVoteDescriptionStyle,
  ProfileHasVotedStyle,
} from './Styled';

type Props = {
  /** Object with all organisation's properties */
  organisation: TypeOrganisation,
  /** Object with all proposal's properties */
  proposal: TypeProposal,
  /** Nature of the organisation's vote */
  voteKey: string,
  /** Proposal's position in list */
  position: number,
  /** Size of proposals list */
  size: number,
};

export const ProfileVoteCard = ({
  voteKey,
  organisation,
  proposal,
  position,
  size,
}: Props) => {
  const { author, question } = proposal;
  const voteAttributes = voteStaticParams[voteKey];
  return (
    <ProfileVoteWrapperStyle aria-posinset={position} aria-setsize={size}>
      <ProfileVoteTitleStyle>
        <ProfileHasVotedStyle
          aria-label={voteAttributes.label}
          color={voteAttributes.color}
        >
          <ButtonIconWrapperStyle transform={voteAttributes.transform}>
            <SvgThumbsUp aria-hidden />
          </ButtonIconWrapperStyle>
        </ProfileHasVotedStyle>
        <ProfileVoteDescriptionStyle>
          <RedLinkStyle
            href={getOrganisationProfileLink(
              organisation.country,
              organisation.language,
              organisation.slug
            )}
          >
            {organisation.organisationName}
          </RedLinkStyle>
          &nbsp;
          <SvgCheckedSymbol
            style={{ fontSize: '14px', fill: TextColors.Blue }}
          />
          &nbsp;
          {i18n.t(`profile.organisation.proposal_${voteKey}`)}
        </ProfileVoteDescriptionStyle>
      </ProfileVoteTitleStyle>

      <ProfileVoteCardStyle>
        <ProposalCardStyle as="div">
          <AuthorWrapperStyle>
            <ProposalAuthorElement
              author={author}
              country={proposal.country}
              language={proposal.language}
              createdAt={proposal.createdAt}
              withAvatar
            />
          </AuthorWrapperStyle>
          <ProposalInnerStyle>
            <ProposalStyle
              id={`proposal_content_${position}`}
              href={getProposalLink(
                proposal.country,
                proposal.language,
                question.slug,
                proposal.id,
                proposal.slug
              )}
            >
              {proposal.content}
            </ProposalStyle>
            <VoteResultElement
              proposalId={proposal.id}
              votes={proposal.votes}
              votedKey={voteKey}
            />
          </ProposalInnerStyle>

          <ProposalFooterWithTagElement tags={proposal.tags} />
        </ProposalCardStyle>
      </ProfileVoteCardStyle>
    </ProfileVoteWrapperStyle>
  );
};
