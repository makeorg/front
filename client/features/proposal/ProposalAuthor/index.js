import * as React from 'react';
import { i18n } from 'Shared/i18n';
import { type Author } from 'Shared/types/proposal';
import { DateHelper } from 'Shared/helpers/date';
import { AuthorInfosStyle, AuthorSeparatorStyle } from './Styled';

type Props = {
  /** Object with author's properties */
  author: Author,
  createdAt?: string,
};

const ProposalAuthorAge = ({ age }) => {
  if (!age) {
    return null;
  }

  return (
    <React.Fragment>
      <span>,&nbsp;</span>
      <span>{i18n.t('proposal_card.author.age', { age })}</span>
    </React.Fragment>
  );
};

export const ProposalAuthor = (props: Props) => {
  const { author, createdAt } = props;

  return (
    <AuthorInfosStyle>
      {author.firstName || author.organisationName}
      <ProposalAuthorAge age={author.age} />
      {!!createdAt && (
        <React.Fragment>
          <AuthorSeparatorStyle aria-hidden>&bull;</AuthorSeparatorStyle>
          <span>{DateHelper.proposalCreationDateFormat(createdAt)}</span>
        </React.Fragment>
      )}
    </AuthorInfosStyle>
  );
};
