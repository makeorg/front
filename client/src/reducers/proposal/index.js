/* @flow */

import * as actionTypes from 'Constants/actionTypes';
import { getBaitText } from 'Constants/proposal';

const initialState: Object = {
  isTyping: false,
  canSubmit: false,
  hasProposed: false,
  isCurrentSubmitSuccess: false,
  content: '',
  length: getBaitText().length,
  questionId: null,
  error: null
};

export default function proposal(state: Object = initialState, action: Object) {
  switch (action.type) {
    case actionTypes.PROPOSE_TYPING:
      return {
        ...state,
        isTyping: true,
        isCurrentSubmitSuccess: false,
        content: action.content,
        length: action.length,
        canSubmit: action.canSubmit
      };
    case actionTypes.PROPOSE_REQUEST:
      return {
        ...state,
        isTyping: false,
        questionId: action.questionId
      };
    case actionTypes.PROPOSE_SUCCESS:
      return {
        ...state,
        isTyping: false,
        canSubmit: false,
        isCurrentSubmitSuccess: true,
        content: '',
        length: getBaitText().length,
        hasProposed: true,
        questionId: null,
        error: null
      };
    case actionTypes.PROPOSE_FAILURE:
      return {
        ...state,
        error: action.error
      };
    case actionTypes.PROPOSAL_LOAD:
      return {
        ...state,
        proposal: action.proposal
      };
    default:
      return state;
  }
}
