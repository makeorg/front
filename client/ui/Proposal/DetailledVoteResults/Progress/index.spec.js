import React from 'react';
import renderer from 'react-test-renderer';
import { proposalTypeFixture } from 'Shared/types/__fixtures__/proposal.fixture';
import { VoteProgress } from './index';

jest.mock('Client/ui/Elements/AccessibilityElements', () => ({
  ScreenReaderItemStyle: 'ScreenReaderItemStyle',
}));

jest.mock('../Styled/Progress', () => ({
  VoteProgressContainerStyle: 'VoteProgressContainerStyle',
  VoteProgressWrapperStyle: 'VoteProgressWrapperStyle',
  VoteCounterStyle: 'VoteCounterStyle',
  VoteProgressItemStyle: 'VoteProgressItemStyle',
}));

describe('VoteProgress', () => {
  const defaultProps = {
    votes: proposalTypeFixture.votes,
    proposalId: proposalTypeFixture.id,
  };

  it('must match the snapshot with default Props', () => {
    const component = renderer
      .create(<VoteProgress {...defaultProps} />)
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
