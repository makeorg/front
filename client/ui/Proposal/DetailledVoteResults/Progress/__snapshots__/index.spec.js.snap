// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoteProgress must match the snapshot with default Props 1`] = `
<VoteProgressContainerStyle>
  <VoteCounterStyle>
    <ScreenReaderItemStyle
      as="span"
    >
      results.static_total
    </ScreenReaderItemStyle>
    vote.label
  </VoteCounterStyle>
  <VoteProgressWrapperStyle>
    <VoteProgressItemStyle
      color="rgb(85,131,34)"
      percent={60}
    />
    <VoteProgressItemStyle
      color="rgb(190,24,44)"
      percent={17}
    />
    <VoteProgressItemStyle
      color="rgb(117,117,117)"
      percent={23}
    />
  </VoteProgressWrapperStyle>
</VoteProgressContainerStyle>
`;
